Add HEALTHCHECK to Dockerfile.

dijit opened the pull request:
The command `wget` is included in the current `alpine:3.13` image;

The `--tries=1` is required because a non-HTTP 200 will cause wget to retry indefinitely.

`|| exit 1` because we want to convert any non-zero exit code to a `1` because that is what docker expects.

This change is needed because GKE Healthchecks on Ingress' are derived entirely from the `HEALTHCHECK` in the Docker image.

dijit committed:
Add HEALTHCHECK to Dockerfile.

The command `wget` is included in the current `alpine:3.13` image;

The --tries=1 is required because a non-HTTP 200 will cause wget to retry indefinitely.

`|| exit 1` because we want to convert any non-zero exit code to a `1` because that is what docker expects.

This change is needed because GKE Healthchecks on Ingress' are derived entirely from the `HEALTHCHECK` in the Docker image.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=33382) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=33382) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=33382) it.</sub>

aknuds1 reviewed (changes requested):
Any healthcheck should also be added to Dockerfile.ubuntu, packaging/docker/*Dockerfile and packaging/docker/custom/*Dockerfile.

dijit commented:
Oh, that one can be nicer as ubuntu packages curl, will add the healthcheck to that also.

dijit committed:
Add healthcheck to Dockerfile.ubuntu

Similar to the alpine version except we can use `curl`

Still keep the `|| exit 1` to sanitise exit codes.

`--fail` simply strips output, which we don't need, we only want the exit code.

aknuds1 commented:
You will need to sign the CLA and also change the other Dockerfiles I mentioned.

dijit committed:
Update ubuntu.Dockerfile

dijit committed:
Update Dockerfile

dijit commented:
the Dockerfiles under `grafana/packaging/docker/custom/` are inheriting from grafana/grafana{,-ubuntu} so it will be inherited automatically.

aknuds1 commented:
> the Dockerfiles under `grafana/packaging/docker/custom/` are inheriting from grafana/grafana{,-ubuntu} so it will be inherited automatically.

Aha, thanks I didn't recall that.

hairyhenderson commented:
> This change is needed because GKE Healthchecks on Ingress' are derived entirely from the `HEALTHCHECK` in the Docker image.

@dijit can you link any supporting documentation for this? I'm no GKE expert, but in general Kubernetes uses the pod’s `readinessProbe` to determine whether to route to a pod.

`HEALTHCHECK` is unique to the Dockerfile format, and [not present in the OCI spec](https://github.com/opencontainers/image-spec/issues/749), so non-Docker container runtimes used in Kubernetes (e.g. containerd/CRI-O/gVisor/etc...) willy simply ignore the `HEALTHCHECK` instruction.

The only place the `HEALTHCHECK` instruction is _not_ ignored to my knowledge is Docker Swarm. It's also worth noting that Official Docker images are explicitly disallowed from including `HEALTHCHECK`: https://github.com/docker-library/faq#healthcheck

dijit commented:
@hairyhenderson I'm mostly going from observed behaviour.

I have the terraform HCL script to create a k8s deployment, service and ingress like so:
```hcl
resource "kubernetes_deployment" "grafana" {

  depends_on = [
    kubernetes_secret.grafana_admin_password,
    kubernetes_secret.grafanadb_connectionstring,
    google_sql_database.grafana_database,
    module.metrics,
    module.logging
  ]

  metadata {
    name      = local.grafana_name
    namespace = kubernetes_namespace.sre_namespace.metadata.0.name
    labels = {
      type = local.grafana_name
    }
  }

  spec {
    replicas = 1
    selector {
      match_labels = {
        role = local.grafana_name
      }
    }

    template {
      metadata {
        labels = {
          namespace = kubernetes_namespace.sre_namespace.metadata.0.name
          type      = local.grafana_name
          role      = local.grafana_name
        }
      }

      spec {
        container {
          image = local.grafana_image
          name  = local.grafana_image_name

          env {
            name  = "GF_USERS_ALLOW_SIGN_UP"
            value = false
          }

          env_from {
            secret_ref {
              name = "${local.grafana_name}-admin-password"
            }
          }

          env_from {
            secret_ref {
              name = "${local.grafana_name}db-connectionstring"
            }
          }

          port {
            container_port = local.grafana_internal_port
            protocol       = "TCP"
          }

          resources {
            limits = {
              cpu    = "750m"
              memory = "1Gi"
            }
            requests = {
              cpu    = "100m"
              memory = "256Mi"
            }
          }

          liveness_probe {
            http_get {
              path = "/api/health"
              port = local.grafana_internal_port
            }

            initial_delay_seconds = 40
            period_seconds        = 10
          }

          readiness_probe {
            http_get {
              path = "/api/health"
              port = local.grafana_internal_port
            }

            initial_delay_seconds = 3
            period_seconds        = 10
          }

          volume_mount {
            name       = "${local.grafana_name}-configmap"
            mount_path = "/etc/grafana/provisioning/datasources"
          }
        }

        volume {
          name = "grafana-configmap"
          config_map {
            name = "grafana-datasource"
          }
        }
      }
    }
  }
}

resource "kubernetes_service" "grafana" {
  metadata {
    name      = local.grafana_name
    namespace = kubernetes_namespace.sre_namespace.metadata.0.name
  }
  spec {
    selector = {
      role = local.grafana_name
    }
    session_affinity = "None"
    port {
      port        = local.grafana_external_port
      target_port = local.grafana_internal_port
      protocol    = "TCP"
    }
    type = "NodePort"
  }
}

resource "kubernetes_ingress" "grafana_ingress" {
  metadata {
    name      = "${local.grafana_name}-ingress"
    namespace = kubernetes_namespace.sre_namespace.metadata.0.name
    labels = {
      type = local.grafana_name
    }
    annotations = {
      "kubernetes.io/ingress.allow-http"          = "false"
      "ingress.gcp.kubernetes.io/pre-shared-cert" = google_compute_managed_ssl_certificate.grafana.name
      delivery                                    = "terraform"
    }
  }
  wait_for_load_balancer = true

  spec {
    rule {
      host = local.grafana_full_host_name
      http {
        path {
          path = "/*"
          backend {
            service_name = local.grafana_name
            service_port = local.grafana_external_port
          }
        }
      }
    }
  }
}
```

Which creates a failing GCE Load Balancer healthcheck which hits `/`

When I shim my healthcheck into the container with another layer it applies it correctly.

aknuds1 commented:
Thanks @hairyhenderson for pointing out that the `HEALTHCHECK` directive isn't generally useful, and ignored in the k8s context. From reading the GKE ingress documentation on [health checks](https://cloud.google.com/kubernetes-engine/docs/concepts/ingress#health_checks) there's also no mention of it, rather other advice that I suggest the PR author consult.

Closing the PR since the GKE argument for adding the `HEALTHCHECK` directive seems to be incorrect.
